hcl2template/types.required_plugins.go

```diff
@@ -41,24 +42,126 @@ func (cfg *PackerConfig) decodeRequiredPluginsBlock(f *hcl.File) hcl.Diagnostics
 }
 
 func (cfg *PackerConfig) decodeImplicitRequiredPluginsBlocks(f *hcl.File) hcl.Diagnostics {
-	// when a plugin is used but not defined in the required plugin blocks, it
-	// is 'implicitly used'. Here we read common configuration blocks to try to
-	// guess plugins.
+	// when a plugin is used but not available it should be 'implicitly
+	// required'. Here we read common configuration blocks to try to guess
+	// plugin usages.
+
+	// decodeRequiredPluginsBlock needs to be called before
+	// decodeImplicitRequiredPluginsBlocks; otherwise all required plugins will
+	// be implicitly required too.
 
 	var diags hcl.Diagnostics
 
 	content, moreDiags := f.Body.Content(configSchema)
 	diags = append(diags, moreDiags...)
 
 	for _, block := range content.Blocks {
+
 		switch block.Type {
 		case sourceLabel:
-			// TODO
+			diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(Builder, block)...)
+		case dataSourceLabel:
+			diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(Datasource, block)...)
+		case buildLabel:
+			content, _, moreDiags := block.Body.PartialContent(buildSchema)
+			diags = append(diags, moreDiags...)
+			for _, block := range content.Blocks {
+
+				switch block.Type {
+				case buildProvisionerLabel:
+					diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(Provisioner, block)...)
+				case buildPostProcessorLabel:
+					diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(PostProcessor, block)...)
+				case buildPostProcessorsLabel:
+					content, _, moreDiags := block.Body.PartialContent(postProcessorsSchema)
+					diags = append(diags, moreDiags...)
+					for _, block := range content.Blocks {
+
+						switch block.Type {
+						case buildPostProcessorLabel:
+							diags = append(diags, cfg.decodeImplicitRequiredPluginsBlock(PostProcessor, block)...)
+						}
+					}
+				}
+			}
+
 		}
 	}
 	return diags
 }
 
+func (cfg *PackerConfig) decodeImplicitRequiredPluginsBlock(k ComponentKind, block *hcl.Block) hcl.Diagnostics {
+	if len(block.Labels) == 0 {
+		// malformed block ? Let's not panic :)
+		return nil
+	}
+	// Currently all block types are `type "component-kind" ["name"] {`
+	// this makes this simple.
+	componentName := block.Labels[0]
+
+	redirect := map[ComponentKind]map[string]string{
+		Builder:       cfg.parser.PluginConfig.BuilderRedirects,
+		PostProcessor: cfg.parser.PluginConfig.PostProcessorRedirects,
+		Provisioner:   cfg.parser.PluginConfig.ProvisionerRedirects,
+		Datasource:    cfg.parser.PluginConfig.DatasourceRedirects,
+	}[k][componentName]
+
+	if redirect == "" {
+		// no known redirect for this component
+		return nil
+	}
+
+	redirectAddr, diags := addrs.ParsePluginSourceString(redirect)
+	if diags.HasErrors() {
+		// This should never happen, since the map is manually filled.
+		return diags
+	}
+
+	for _, req := range cfg.Packer.RequiredPlugins {
+		if _, found := req.RequiredPlugins[redirectAddr.Type]; found {
```

Comment:
Nice! I didn't think about this specific case.

Reply:
Nice catch